File: Windows/src/main/remoteConnect/LanSignalingResolver.ts
import type { SignalingClient } from "../signaling/SignalingClient.js";

export type LanDirectConfig = {
  host: string;
  port: number;
  token: string;
  transport: "lan";
};

export async function resolveLanViaSignaling(
  signalingClient: SignalingClient,
  connectionId: number,
  targetDeviceId: number,
  timeoutMs = 15_000
): Promise<LanDirectConfig | null> {
  if (signalingClient.status !== "connected") {
    return null;
  }

  return new Promise((resolve) => {
    let finished = false;
    const finish = (config: LanDirectConfig | null) => {
      if (finished) return;
      finished = true;
      clearTimeout(timer);
      signalingClient.removeTunnelHandler(connectionId);
      signalingClient.sendTunnelClose(connectionId, "lan_probe_done");
      resolve(config);
    };

    signalingClient.setTunnelHandler(connectionId, (event) => {
      switch (event.type) {
        case "tunnel_open_ack":
          signalingClient.sendTunnelFrame(connectionId, Date.now(), JSON.stringify({ type: "lan_request" }));
          break;
        case "tunnel_frame": {
          const config = parseLanOffer(event.frame);
          if (config) finish(config);
          break;
        }
        case "tunnel_close":
        case "tunnel_error":
          finish(null);
          break;
        default:
          break;
      }
    });

    if (!signalingClient.openTunnel(connectionId, targetDeviceId)) {
      finish(null);
      return;
    }
    signalingClient.sendTunnelFrame(connectionId, 1, JSON.stringify({ type: "lan_request" }));
    const timer = setTimeout(() => finish(null), timeoutMs);
  });
}

function parseLanOffer(frame?: string): LanDirectConfig | null {
  if (!frame) return null;
  try {
    const json = JSON.parse(frame) as { type?: string; ip?: string; port?: number; token?: string };
    if (json.type !== "lan_offer") return null;
    const ip = json.ip?.trim() ?? "";
    const port = json.port ?? 0;
    const token = json.token?.trim() ?? "";
    if (!ip || port < 1 || port > 65_535 || !token) return null;
    return { host: ip, port, token, transport: "lan" };
  } catch {
    return null;
  }
}

File: Android/app/src/main/java/com/acode/android/data/LanSignalingResolver.kt
package com.acode.android.data

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import org.json.JSONObject

/**
 * When the cloud lan-token API is unavailable, negotiate LAN endpoint + transient token
 * over the signaling tunnel immediately after connect is accepted.
 */
object LanSignalingResolver {
    suspend fun resolve(
        connectionId: Int,
        targetDeviceId: Int,
        signalingClient: RemoteSignalingClient,
        timeoutMs: Long = 15_000,
    ): RemoteChatConfig? = withContext(Dispatchers.IO) {
        if (!signalingClient.isConnected) return@withContext null

        val offer = CompletableDeferred<RemoteChatConfig?>()
        val previousTunnelHandler = signalingClient.onTunnelEvent

        signalingClient.onTunnelEvent = { type, eventConnectionId, frame, reason ->
            if (eventConnectionId == connectionId) {
                when (type) {
                    "tunnel_open_ack" -> {
                        signalingClient.sendTunnelFrame(
                            connectionId,
                            System.nanoTime(),
                            JSONObject().put("type", "lan_request").toString(),
                        )
                    }
                    "tunnel_frame" -> {
                        parseLanOffer(frame)?.let { config ->
                            if (!offer.isCompleted) offer.complete(config)
                        }
                    }
                }
            } else {
                previousTunnelHandler?.invoke(type, eventConnectionId, frame, reason)
            }
        }

        val opened = signalingClient.openTunnel(connectionId, targetDeviceId)
        if (!opened) {
            signalingClient.onTunnelEvent = previousTunnelHandler
            return@withContext null
        }
        signalingClient.sendTunnelFrame(
            connectionId,
            1,
            JSONObject().put("type", "lan_request").toString(),
        )

        try {
            withTimeoutOrNull(timeoutMs) { offer.await() }
        } finally {
            signalingClient.sendTunnelClose(connectionId, "lan_probe_done")
            signalingClient.onTunnelEvent = previousTunnelHandler
        }
    }

    private fun parseLanOffer(frame: String?): RemoteChatConfig? {
        val json = runCatching { JSONObject(frame.orEmpty()) }.getOrNull() ?: return null
        if (json.optString("type") != "lan_offer") return null
        val ip = json.optString("ip").trim()
        val port = json.optInt("port")
        val token = json.optString("token").trim()
        if (ip.isBlank() || port !in 1..65535 || token.isBlank()) return null
        return RemoteChatConfig(macHost = ip, port = port, token = token, transport = "lan")
    }
}
